lint_rules/lint_rules/ifs_arpege_coding_standards.py

```diff
@@ -0,0 +1,196 @@
+)
+from loki import ir
+
+
+__all__ = [
+    'MissingImplicitNoneRule', 'OnlyParameterGlobalVarRule', 'MissingIntfbRule',
+]
+
+
+class MissingImplicitNoneRule(GenericRule):
+    """
+    ``IMPLICIT NONE`` must be present in all scoping units but may be omitted
+    in module procedures.
+    """
+
+    type = RuleType.SERIOUS
+
+    docs = {
+        'id': 'L1',
+        'title': (
+            'IMPLICIT NONE must figure in all scoping units. '
+            'Once per module is sufficient.'
+        ),
+    }
+
+    _regex = re.compile(r'implicit\s+none\b', re.I)
+
+    @classmethod
+    def check_for_implicit_none(cls, ir_):
+        """
+        Check for intrinsic nodes that match the regex.
+        """
+        for intr in FindNodes(ir.Intrinsic).visit(ir_):
+            if cls._regex.match(intr.text):
+                break
+        else:
+            return False
+        return True
+
+    @classmethod
+    def check_module(cls, module, rule_report, config):
+        """
+        Check for ``IMPLICIT NONE`` in the module's spec.
+        """
+        found_implicit_none = cls.check_for_implicit_none(module.spec)
+        if not found_implicit_none:
+            # No 'IMPLICIT NONE' intrinsic node was found
+            rule_report.add('No `IMPLICIT NONE` found', module)
+
+    @classmethod
+    def check_subroutine(cls, subroutine, rule_report, config, **kwargs):
+        """
+        Check for ``IMPLICIT NONE`` in the subroutine's spec or an enclosing
+        :any:`Module` scope.
+        """
+        found_implicit_none = cls.check_for_implicit_none(subroutine.ir)
+
+        # Check if enclosing scopes contain implicit none
+        scope = subroutine.parent
+        while scope and not found_implicit_none:
+            if isinstance(scope, Module) and hasattr(scope, 'spec') and scope.spec:
+                found_implicit_none = cls.check_for_implicit_none(scope.spec)
+            scope = scope.parent if hasattr(scope, 'parent') else None
+
+        if not found_implicit_none:
+            # No 'IMPLICIT NONE' intrinsic node was found
+            rule_report.add('No `IMPLICIT NONE` found', subroutine)
+
+
+class OnlyParameterGlobalVarRule(GenericRule):
+    """
+    Only parameters to be declared as global variables.
+    """
+
+    type = RuleType.SERIOUS
+
+    docs = {
+        'id': 'L3',
+        'title': 'Only parameters to be declared as global variables.'
+    }
+
+    @classmethod
+    def check_module(cls, module, rule_report, config):
+        for decl in module.declarations:
+            if not decl.symbols[0].type.parameter:
+                msg = f'Global variable(s) declared that are not parameters: {", ".join(s.name for s in decl.symbols)}'
+                rule_report.add(msg, decl)
+
+
+class MissingIntfbRule(GenericRule):
+    """
+    Calls to subroutines and functions that are provided neither by a module
+    nor by a CONTAINS statement, must have a matching explicit interface block.
+    """
```

Comment:
No action: For practicality this rule might need to be able to handle "exceptions", like for example "gstats" to make the output more usable. But this is up for debate and could be added later.

Reply:
The coding standard rule does not currently allow for any exceptions: https://sites.ecmwf.int/docs/ifs-arpege-coding-standards/fortran/rules/L9.html
For gstats in particular, the fix should likely be the addition of the interface block - it's peculiar that this was missing in the first place and I don't see any advantage in omitting it.

From a practical point of view I agree, this would be something that can be easily added to the rule.